### PR TITLE
Enhancement: Call jls command only once for lists of jails

### PR DESCRIPTION
**Pull request by gronke**
- adds JailState data structure
- the Jails list iterator invokes update of jails state with jls only once and passes fitting results to jails as it iterates them

A quick comparison with 10 jails on SSD:
- Before: 2.51s real
- After: 0.6s real

### CODE REVIEW

File: iocage/cli/start.py [hunk 1]
```diff
@@ -98,6 +98,7 @@ def start_jails(jails, logger, print_function):
             jail.require_jail_not_template()
             print_function(jail.start())
         except iocage.lib.errors.IocageException:
+            raise
             failed_jails.append(jail)
```
Comment (igalic) [inline iocage/cli/start.py]: how's this supposed to work?

File: iocage/cli/stop.py [hunk 2]
```diff
@@ -86,7 +86,7 @@ def stop_jails(
     for jail in jails:
         try:
             print_function(jail.stop(force=force))
-        except iocage.lib.errors.IocageException:
+        except iocage.lib.errors.IocageException as e:
```
Comment (igalic) [inline iocage/cli/stop.py]: you're not using `e`

File: iocage/lib/Jail.py [hunk 3]
```diff
@@ -450,7 +462,7 @@ def rename(self, new_name: str) -> None:
             self.logger.verbose(
                 f"Dataset {current_dataset_name} renamed to {new_dataset_name}"
             )
-        except:
+        except BaseException:
```
Comment (igalic) [inline iocage/lib/Jail.py]: what's `BaseException`?

File: iocage/lib/JailState.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        self,
+        states: typing.Optional[JailStatesDict]=None
+    ) -> None:
+
+        if states is None:
+            dict.__init__(self, {})
+        else:
+            dict.__init__(self, states)
+
+    def query(self):
```
Comment (igalic) [inline iocage/lib/JailState.py]: is this a standard `dict` function here?

File: iocage/lib/Jails.py [hunk 5]
```diff
@@ -76,8 +77,32 @@ def _create_resource_instance(
...
+            self.logger.spam(
+                f"Injecting pre-loaded state to '{jail.humanreadable_name}'"
+            )
+            jail.jail_state = self.states[jail.identifier]
+
         return jail
 
+    def __iter__(
+        self
+    ) -> typing.Generator['iocage.lib.Resource.Resource', None, None]:
```
Comment (igalic) [inline iocage/lib/Jails.py]: is this necessary to be a `'ForwardReference'`?

File: iocage/lib/Jail.py (hunk 3 above)
Comment (gronke) [reply iocage/lib/Jail.py]: All Python Exceptions inherit from BaseException.

File: iocage/lib/JailState.py (hunk 4 above)
Comment (gronke) [reply iocage/lib/JailState.py]: no, but update (my previous naming choice) was. This is called to query the state from jls.

File: iocage/lib/Jails.py (hunk 5 above)
Comment (gronke) [reply iocage/lib/Jails.py]: No, I don't see the requirement, but passed it as string for convenience with other situations where it makes a difference. (Idea for the future: We could aim to refactor all circular dependencies to a linear dependency tree.)

**Pull request merged**